src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Release a stream with specified stream ID.
+    pub fn pcm_release(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmRelease);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Start a stream with specified stream ID.
+    pub fn pcm_start(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStart);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Stop a stream with specified stream ID.
+    pub fn pcm_stop(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStop);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Transfer PCM frame to device, based on the stream type(OUTPUT/INPUT).
+    /// 
+    /// This is a blocking method that will not return until the audio playback is complete.
+    pub fn pcm_xfer(&mut self, stream_id: u32, frames: &[u8]) -> Result {
+        const U32_SIZE: usize = mem::size_of::<u32>();
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if !self.pcm_parameters[stream_id as usize].setup {
+            warn!("Please set parameters for a stream before using it!");
+            return Err(Error::IoError);
+        }
+        let buffer_size = self.pcm_parameters[stream_id as usize].buffer_bytes as usize;
+        assert!(buffer_size * 2 <= frames.len());
```

Comment:
Why is the caller not allowed to send a smaller amount of data?

Reply:
Now I've fixed this bug, and now caller is allowed to send a smaller data.